services/replicator/replicator.go

```diff
@@ -258,10 +257,8 @@ func (r *Replicator) OpenReplicationReadStreamHandler(w http.ResponseWriter, req
 	inConn := newInConnection(extUUID, destDesc.GetPath(), inStream, outConn.msgsCh, r.logger, r.m3Client, metrics.OpenReplicationReadScope, metrics.OpenReplicationReadPerDestScope)
 	inConn.open()
 
-	go r.manageInOutConn(inConn, outConn)
-	<-inConn.closeChannel
-	<-outConn.closeChannel
-
+	outConn.WaitUntilDone()
+	inConn.WaitUntilDone()
```

Comment:
i might suggest setting up a waitgroup within this function .. and passing a pointer to the waitgroup to inconn/outconn, etc that they increment/decrement .. and you can just do a wg.Wait() here that will automatically wait for both, etc.